rts/motoko-rts-tests/src/gc/heap.rs

```diff
@@ -289,13 +315,38 @@ fn create_dynamic_heap(
     for (obj, refs) in refs {
         let obj_offset = object_addrs.get(obj).unwrap() - heap_start;
         for (ref_idx, ref_) in refs.iter().enumerate() {
-            // -1 for skewing
             let ref_addr = make_pointer(*object_addrs.get(ref_).unwrap() as u32);
-            let field_offset = obj_offset + (3 + ref_idx) * WORD_SIZE;
+            let field_offset = obj_offset
+                + (size_of::<Array>() + Words(1 + ref_idx as u32))
+                    .to_bytes()
+                    .0 as usize;
             write_word(dynamic_heap, field_offset, u32::try_from(ref_addr).unwrap());
         }
     }
 
+    // Add the closure table
+    let n_objects = refs.len();
+    // fields+1 for the scalar field (idx)
+    let n_fields: usize = refs.values().map(|fields| fields.len() + 1).sum();
+    let closure_table_offset =
+        (size_of::<Array>() * n_objects as u32).to_bytes().0 as usize + n_fields * WORD_SIZE;
+
+    {
+        let mut heap_offset = closure_table_offset;
+
+        write_word(dynamic_heap, closure_table_offset, TAG_ARRAY);
+        heap_offset += WORD_SIZE;
+
+        write_word(dynamic_heap, heap_offset, closure_table.len() as u32);
+        heap_offset += WORD_SIZE;
+
+        for idx in closure_table {
+            let idx_ptr = *object_addrs.get(idx).unwrap();
+            write_word(dynamic_heap, heap_offset, make_pointer(idx_ptr as u32));
+            heap_offset += WORD_SIZE;
+        }
```

Comment:
This assumes there are no unused slots at the end of the closure table right? Ie. all the fields in the array will have sensible motoko values after this setup?

Reply:
Correct. I will generalize this as needed.